test/broadcasting.jl

```diff
@@ -0,0 +1,364 @@
+    @test df.x1 == [2, 2, 2]
+    @test df.x2 == [2, 2, 2]
+    @test df[3:end] == refdf[3:end]
+
+    dfv = @view df[1:2, 3:end]
+    dfv[:, [:x3,:x4]] .= Matrix(dfv[[:x3,:x4]]) .+ 1
+    @test dfv.x3 == [2, 2]
+    @test dfv.x4 == [2, 2]
+    @test dfv[3:end] == refdf[1:2, 5:end]
+    @test all(Matrix(df[1:2, 1:2]) .== 2)
+
+    df = copy(refdf)
+    df[[:x1,:x2]] .= Matrix(df[[:x1,:x2]]) .+ [1 1
+                                               1 1
+                                               1 1]
+    @test df.x1 == [2, 2, 2]
+    @test df.x2 == [2, 2, 2]
+    @test df[3:end] == refdf[3:end]
+
+    dfv = @view df[1:2, 3:end]
+    dfv[[:x3,:x4]] .= Matrix(dfv[[:x3,:x4]]) .+ [1 1
+                                                 1 1]
+    @test dfv.x3 == [2, 2]
+    @test dfv.x4 == [2, 2]
+    @test dfv[3:end] == refdf[1:2, 5:end]
+    @test all(Matrix(df[1:2, 1:2]) .== 2)
+
+    df = copy(refdf)
+    df[:, [:x1,:x2]] .= Matrix(df[[:x1,:x2]]) .+ [1 1
+                                                  1 1
+                                                  1 1]
+    @test df.x1 == [2, 2, 2]
+    @test df.x2 == [2, 2, 2]
+    @test df[3:end] == refdf[3:end]
+
+    dfv = @view df[1:2, 3:end]
+    dfv[:, [:x3,:x4]] .= Matrix(dfv[[:x3,:x4]]) .+ [1 1
+                                                    1 1]
+    @test dfv.x3 == [2, 2]
+    @test dfv.x4 == [2, 2]
+    @test dfv[3:end] == refdf[1:2, 5:end]
+    @test Matrix(df[1:2, 1:2]) == fill(2, (2, 2))
+
+    df = copy(refdf)
+    dfv = @view df[1:2, 2:end]
+    @test_throws DimensionMismatch df[[:x1,:x2]] .= rand(3, 10)
+    @test_throws DimensionMismatch dfv[[:x3,:x4]] .= rand(2, 10)
+    @test_throws DimensionMismatch df[:, [:x1,:x2]] .= rand(3, 10)
+    @test_throws DimensionMismatch dfv[:, [:x3,:x4]] .= rand(2, 10)
+end
+
+@testset "assignment to a whole data frame and data frame row" begin
+    df = copy(refdf)
+    df .= 10
+    @test all(Matrix(df) .== 10)
+    dfv = view(df, 1:2, 1:4)
+    dfv .= 100
+    @test (all(Matrix(df[1:2, 1:4]) .== 100))
+    @test (all(Vector(df[3, 1:4]) .== 10))
+    dfr = df[1, 1:2]
+    dfr .= 1000
+    @test (all(Vector(df[1, 1:2]) .== 1000))
+    @test (all(Vector(df[3, :]) .!= 1000))
+
+    df = copy(refdf)
+    df[:] .= 10
+    @test all(Matrix(df) .== 10)
+    dfv = view(df, 1:2, 1:4)
+    dfv[:] .= 100
+    @test (all(Matrix(df[1:2, 1:4]) .== 100))
+    @test (all(Vector(df[3, 1:4]) .== 10))
+    dfr = df[1, 1:2]
+    dfr[:] .= 1000
+    @test (all(Vector(df[1, 1:2]) .== 1000))
+    @test (all(Vector(df[3, :]) .!= 1000))
+
+    df = copy(refdf)
+    df[:,:] .= 10
+    @test all(Matrix(df) .== 10)
+    dfv = view(df, 1:2, 1:4)
+    dfv[:, :] .= 100
+    @test (all(Matrix(df[1:2, 1:4]) .== 100))
+    @test (all(Vector(df[3, 1:4]) .== 10))
+end
+
+@testset "extending data frame in broadcasted assignment - one column" begin
+    df = copy(refdf)
+    df[:a] .= 1
+    @test all(Matrix(df) .== 1)
+    @test names(df)[end] == :a
```

Comment:
Also test that `df[1:end-1] == refdf` just in case (here and below).

Reply:
fixed